src/gt4py/next/type_system/type_translation.py

```diff
@@ -188,18 +199,23 @@ def from_value(value: Any) -> ts.TypeSpec:
     elif common.is_field(value):
         dims = list(value.domain.dims)
         dtype = from_type_hint(value.dtype.scalar_type)
+        assert isinstance(dtype, ts.ScalarType)
         symbol_type = ts.FieldType(dims=dims, dtype=dtype)
     elif isinstance(value, tuple):
         # Since the elements of the tuple might be one of the special cases
         # above, we can not resort to generic `infer_type` but need to do it
         # manually here. If we get rid of all the special cases this is
         # not needed anymore.
-        return ts.TupleType(types=[from_value(el) for el in value])
+        elems = [from_value(el) for el in value]
+        assert all(isinstance(elem, ts.DataType) for elem in elems)
+        return ts.TupleType(types=elems)  # type: ignore[arg-type] # checked in assert
     else:
         type_ = xtyping.infer_type(value, annotate_callable_kwargs=True)
         symbol_type = from_type_hint(type_)
 
-    if isinstance(symbol_type, (ts.DataType, ts.CallableType, ts.OffsetType, ts.DimensionType)):
+    if isinstance(
+        symbol_type, (ts.DataType, ts.CallableType, ts.OffsetType, ts.DimensionType)
```

Comment:
CallableType is not a TypeSpec. this looks like a bug

Reply:
CallableType is a mixin and all subtypes of it are a TypeSpec. Maybe we should rename it to something like `CallableTypeMixin`. If you want it to be 100% correct maybe `isinstance(symbol_type, ts.CallableType) and isinstance(symbol_type, ts.TypeSpec)`?

Reply:
fixed by `isinstance(symbol_type, ts.CallableType) and isinstance(symbol_type, ts.TypeSpec)`